fsm/fsm.c

```diff
@@ -0,0 +1,206 @@
+        {
+            break;
+        }
+        get_next_state(input, &state);
+        tmp = *state;
+        fflush(stdin);
+        state = &tmp;
+    }
+
+    test_fsm();
+    printf("Unit Test Passed");
+}
+
+// Unit test asserting correct state transitions
+// returns true if pass, false otherwise
+void test_fsm()
+{
+    char input;
+
+    // Test setup
+    State * boot;
+    State boot_val = BOOT;
+    State * autonomous;
+    State autonomous_val = AUTONOMOUS;
+    State * manual;
+    State manual_val = MANUAL;
+
+    // Boot -> Boot transition tests
+    boot = &boot_val;
+    input = 'p';
+    get_next_state(input, &boot);
+    assert(*boot == boot_val);
+
+    boot = &boot_val;
+    input = 'P';
+    get_next_state(input, &boot);
+    assert(*boot == boot_val);
+
+    boot = &boot_val;
+    input = 'w';
+    get_next_state(input, &boot);
+    assert(*boot == boot_val);
+
+    boot = &boot_val;
+    input = 'W';
+    get_next_state(input, &boot);
+    assert(*boot == boot_val);
+
+    // Boot -> Autonomous transition tests
+    boot = &boot_val;
+    input = 'a';
+    get_next_state(input, &boot);
+    assert(*boot == autonomous_val);
+
+    boot = &boot_val;
+    input = 'A';
+    get_next_state(input, &boot);
+    assert(*boot == autonomous_val);
+
+    // Boot -> Manual transition tests
+    boot = &boot_val;
+    input = 'm';
+    get_next_state(input, &boot);
+    assert(*boot == manual_val);
+
+    boot = &boot_val;
+    input = 'M';
+    get_next_state(input, &boot);
+    assert(*boot == manual_val);
+
+    // Autonomous -> Boot transition tests
+    autonomous = &autonomous_val;
+    input = 'p';
+    get_next_state(input, &autonomous);
+    assert(*autonomous == boot_val);
+
+    autonomous = &autonomous_val;
+    input = 'P';
+    get_next_state(input, &autonomous);
+    assert(*autonomous == boot_val);
+
+    autonomous = &autonomous_val;
+    input = 'w';
+    get_next_state(input, &autonomous);
+    assert(*autonomous == boot_val);
+
+    autonomous = &autonomous_val;
+    input = 'W';
+    get_next_state(input, &autonomous);
+    assert(*autonomous == boot_val);
+
+    // Manual -> Boot transition tests
+    manual = &manual_val;
+    input = 'p';
+    get_next_state(input, &manual);
+    assert(*manual == boot_val);
+
+    manual = &manual_val;
+    input = 'P';
+    get_next_state(input, &manual);
+    assert(*manual == boot_val);
+
+    manual = &manual_val;
+    input = 'w';
+    get_next_state(input, &manual);
+    assert(*manual == boot_val);
+
+    manual = &manual_val;
+    input = 'W';
+    get_next_state(input, &manual);
+    assert(*manual == boot_val);
+}
+
+// Get the next state based on the current state and input
+void get_next_state(char input, State **state)
+{
+    State boot = BOOT;
+    State autonomous = AUTONOMOUS;
+    State manual = MANUAL;
+
+    int curr_state = (int)**state;
+
+    if (curr_state == (int)boot)
+    {
+        if (input == 'P' || input == 'p' || input == 'W' || input == 'w')
```

Comment:
this `if` is the same for every state, IMO you should just pull it out to the beginning and break early if called